### PR TITLE
Maps Switch to unique_ptrs

### PR DESCRIPTION
**Pull request by JoshuaSBrown**
Part 3

### CODE REVIEW

File: include/votca/csg/topologymap.h
Comment (JensWehner) [inline include/votca/csg/topologymap.h]: why not just make it

```suggestion
  using MapContainer = std::vector<Map>;
```

Comment (JensWehner) [reply include/votca/csg/topologymap.h]: and make it value based.

**Review by JensWehner — CHANGES_REQUESTED** [review]
either use `make_unique` or preferably make it value based and not pointer based

**Comment by junghans** [comment]
@JoshuaSBrown 
```
/__w/csg/csg/votca/csg/include/votca/csg/topologymap.h:52:33: error: no matching function for call to 'std::vector<votca::csg::Map>::push_back(std::remove_reference<std::unique_ptr<votca::csg::Map>&>::type)'
   52 |   _maps.push_back(std::move(map));
      |                                 ^
```

**Comment by junghans** [comment]
No app using the mapping fail:
```
	 92 - integration_Run_csg_fmatch (SEGFAULT)
	 94 - integration_Run_csg_fmatch_3body (SEGFAULT)
	100 - integration_Run_csg_map_gro (SEGFAULT)
```

**Comment by JoshuaSBrown** [comment]
Might need to wait until the topology map pr gets merged before this will work.

File: include/votca/csg/map.h
Comment (JensWehner) [inline include/votca/csg/map.h]: I do not understand.

**Review by JensWehner — CHANGES_REQUESTED** [review]
I do not understand the hiding of all the constructors. I understand that you want to protect certain conditions, but in my mind that is what assert statements are for. I do not want to couple constructors to other classes, even less friend them, because then encapsulation goes out of the window. 

If you want to create a common interface for object creation, I would make a factory.

**Comment by JoshuaSBrown** [comment]
> I do not understand the hiding of all the constructors. I understand that you want to protect certain conditions, but in my mind that is what assert statements are for. I do not want to couple constructors to other classes, even less friend them, because then encapsulation goes out of the window.
> 
> If you want to create a common interface for object creation, I would make a factory.

Ok, I think I can make that work, I'll just move all the child classes into the source files, that should hide them from being used where they are not supposed to.

**Comment by junghans** [comment]
@votca-bot format

**Review by JensWehner — APPROVED** [review]
thanks

**Pull request merged**
